mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala

```diff
@@ -177,102 +177,108 @@ object GradientBoostedTrees extends Logging {
     treeStrategy.assertValid()
 
     // Cache input
-    if (input.getStorageLevel == StorageLevel.NONE) {
+    val persistedInput = if (input.getStorageLevel == StorageLevel.NONE) {
       input.persist(StorageLevel.MEMORY_AND_DISK)
-    }
+      true
+    } else false
 
-    timer.stop("init")
+    try {
```

Comment:
I too think it makes some sense but have not seen it done elsewhere.  I think I asked about it when I first started working on Spark and was told something like @srowen said.  Even if it's "leaked," other RDDs can still push it out of memory or disk if needed, right?  Have you encountered cases where that does not happen?